src/node_perf.cc

```diff
@@ -529,21 +574,117 @@ bool ELDHistogram::RecordDelta() {
 }
 
 bool ELDHistogram::Enable() {
-  if (enabled_ || IsHandleClosing()) return false;
-  enabled_ = true;
-  prev_ = 0;
-  uv_timer_start(&timer_,
-                 DelayIntervalCallback,
-                 resolution_,
-                 resolution_);
+  if (IsHandleClosing()) {
+    return false;
+  }
+  if (!EventLoopHistogram::Enable()) {
+    return false;
+  }
+  CHECK_EQ(uv_timer_start(
+        &timer_,
+        DelayIntervalCallback,
+        resolution_,
+        resolution_),
+      0);
   uv_unref(reinterpret_cast<uv_handle_t*>(&timer_));
   return true;
 }
 
 bool ELDHistogram::Disable() {
-  if (!enabled_ || IsHandleClosing()) return false;
-  enabled_ = false;
-  uv_timer_stop(&timer_);
+  if (IsHandleClosing()) {
+    return false;
+  }
+  if (!EventLoopHistogram::Disable()) {
+    return false;
+  }
+  CHECK_EQ(uv_timer_stop(&timer_), 0);
+  return true;
+}
+
+ELIHistogram::ELIHistogram(
+    Environment* env,
+    Local<Object> wrap,
+    int32_t resolution)
+  : AsyncWrap(
+      env,
+      wrap,
+      AsyncWrap::PROVIDER_ELIHISTOGRAM),
+    EventLoopHistogram(resolution) {
+  MakeWeak();
+  CHECK_EQ(uv_prepare_init(env->event_loop(), &prepare_), 0);
+  CHECK_EQ(uv_check_init(env->event_loop(), &check_), 0);
+}
+
+ELIHistogram::~ELIHistogram() {
+  // uv_prepare_t and uv_check_t can be closed synchronously
```

Comment:
I don’t think this is right – `uv_close()` calls `uv__make_close_pending` unconditionally, whether there is a callback associated with them or not. In particular, the handles *will* end up as part of the loop’s `closing_handles` linked list after their memory has already been deallocated, so this does open up the possibility of use-after-free accesses…

Reply:
Good point! Sorry I missed this.

I'm somewhat reluctant to do anything complicated here. Do you think it should just allocate the handles and `delete` them in CloseCb?

Reply:
I think that should work, but I think I’d have a mild preference for something else:

My suggestion would be to let this class inherit from `HandleWrap` and pass the `uv_prepare_t` to that, and then for closing the handle, override `HandleWrap::Close(cb)`, e.g. as 

```c++
void ELIHistogram::Close(Local<Value> close_cb) {
  if (!IsHandleClosing()) uv_close(&check_, nullptr);
  // Let HandleWrap take care of the prepare handle
  HandleWrap::Close(close_cb);  
}
```

It’s not the prettiest thing in the world, but I think it should work without causing trouble and probably takes the least amount of extra code?

Reply:
Sounds like a good idea. I wish we had `MultiHandleWrap` :joy:

Reply:
Ack.